crates/red_knot_python_semantic/src/types/intern.rs

```diff
@@ -0,0 +1,71 @@
+use crate::types::{
+    ClassType, FileClassTypeId, FileFunctionTypeId, FileIntersectionTypeId, FileUnionTypeId,
+    FunctionType, IntersectionType, ModuleType, UnionType,
+};
+
+use ruff_index::IndexVec;
+
+/// Interned types for a file
+#[derive(Debug, Eq, PartialEq, Clone)]
+pub(crate) struct FileTypeStore {
+    /// The type of the module.
+    module_type: ModuleType,
+
+    /// The types of the defined classes in this module.
+    class_types: IndexVec<FileClassTypeId, ClassType>,
+
+    /// The types of the defined functions in this module.
+    function_types: IndexVec<FileFunctionTypeId, FunctionType>,
+
+    union_types: IndexVec<FileUnionTypeId, UnionType>,
+    intersection_types: IndexVec<FileIntersectionTypeId, IntersectionType>,
+}
+
+impl FileTypeStore {
+    pub(super) fn new(module_type: ModuleType) -> Self {
+        Self {
+            module_type,
+            class_types: IndexVec::default(),
+            function_types: IndexVec::default(),
+            union_types: IndexVec::default(),
+            intersection_types: IndexVec::default(),
+        }
+    }
+
+    pub(super) fn module_ty(&self) -> &ModuleType {
+        &self.module_type
+    }
+
+    pub(super) fn class_ty(&self, id: FileClassTypeId) -> &ClassType {
```

Comment:
I think returning a refence would work if we use an actual `Arena` that allocates a new vector when it runs out of space instead of resizing the existing one (which would invalidate previously returned Ids). 

Or more generally. The way to return references here is by lifting the lifetime of types outside of type store by having the actual data owned further up.